### PR TITLE
Part three: Album

### PR DESCRIPTION
**Pull request by ltxitai**
Show multiple images as well as promotional (Lightricks) image.
Allow opening images in full screen. Tapping promotional image opens
Lightricks' website.
Images are cached so the same image may be reused if the view scrolls
back to it.

### CODE REVIEW

File: Baygur/BGRCachingImageFetchStrategy.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) 2019 Lightricks. All rights reserved.
+// Created by Itai Shaked.
+
+#import "BGRRemoteImageFetchStrategy.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface BGRCachingImageFetchStrategy : NSObject <BGRRemoteImageFetchStrategy>
```
Comment (gershonho) [inline Baygur/BGRCachingImageFetchStrategy.h]: Doc

File: Baygur/BGRCachingImageFetchStrategy.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+- (instancetype)init NS_UNAVAILABLE;
+
+- (instancetype)initWithFetchStrategy:(id<BGRRemoteImageFetchStrategy>)fetchStrategy
```
Comment (gershonho) [inline Baygur/BGRCachingImageFetchStrategy.h]: Doc

File: Baygur/BGRCachingImageFetchStrategy.mm [hunk 3]
```diff
@@ -0,0 +1,44 @@
+// Copyright (c) 2019 Lightricks. All rights reserved.
+// Created by Itai Shaked.
+
+#import "BGRCachingImageFetchStrategy.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface BGRCachingImageFetchStrategy ()
+@property (readonly, nonatomic) id<BGRRemoteImageFetchStrategy> fetchStrategy;
```
Comment (gershonho) [inline Baygur/BGRCachingImageFetchStrategy.mm]: Doc, add empty lines.

File: Baygur/BGRRemoteImageFetchStrategy.h [hunk 4]
```diff
@@ -0,0 +1,13 @@
+// Copyright (c) 2019 Lightricks. All rights reserved.
+// Created by Itai Shaked.
+
+NS_ASSUME_NONNULL_BEGIN
+
+@protocol BGRRemoteImageFetchStrategy <NSObject>
```
Comment (gershonho) [inline Baygur/BGRRemoteImageFetchStrategy.h]: Doc

File: Baygur/BGRRemoteImageFetchStrategy.h [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+- (void)fetch:(NSURL *)url completionHandler:(void (^)(NSURL *requestURL, UIImage * _Nullable image,
```
Comment (gershonho) [inline Baygur/BGRRemoteImageFetchStrategy.h]: 1. Doc.
2. Define a type for callback, use the type in all implementation classes.

File: Baygur/BGRImageView.mm [hunk 6]
```diff
@@ -13,17 +13,17 @@ @interface BGRImageView () <UIScrollViewDelegate>
...
-
-/// Shown while loading an image. 
-@property (readonly, nonatomic) UIActivityIndicatorView *activityIndicator;
-
-/// Used do retrieve images.
-@property (readonly, nonatomic) NSURLSession *session;
-
-/// Current image download task in progress.
-@property (readwrite, nullable, nonatomic) NSURLSessionTask *currentTask;
+///// Used to display an error message if the image could not be loaded.
```
Comment (gershonho) [inline Baygur/BGRImageView.mm]: Drop all commented-out code

File: Baygur/BGRImageCell.h [hunk 7]
```diff
@@ -0,0 +1,13 @@
+// Copyright (c) 2019 Lightricks. All rights reserved.
+// Created by Itai Shaked.
+
+#import "BGRImageDisplayer.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@class BGRImageMetadata;
+
+@interface BGRImageCell : UICollectionViewCell <BGRImageDisplayer>
```
Comment (gershonho) [inline Baygur/BGRImageCell.h]: doc

File: Baygur/BGRImageCell.mm [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+#import "BGRImageMetadata.h"
+#import "BGRRemoteImageView.h"
+#import "BGRURLSessionImageFetchStrategy.h"
+#import "BGRCachingImageFetchStrategy.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface BGRImageCell ()
+@property (readonly, nonatomic) BGRRemoteImageView *imageView;
```
Comment (gershonho) [inline Baygur/BGRImageCell.mm]: doc

File: Baygur/BGRImageCell.mm [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+@end
+
+@implementation BGRImageCell
+
+- (instancetype)init {
```
Comment (gershonho) [inline Baygur/BGRImageCell.mm]: Drop all unneeded initialisers, mark them `NS_UNAVAILABLE` in the header file.

File: Baygur/BGRImageCell.mm [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+- (nullable instancetype)initWithCoder:(NSCoder *)aDecoder {
+  if (self = [super initWithCoder:aDecoder]) {
+    [self setup];
+  }
+  return self;
+}
+
+static id<BGRRemoteImageFetchStrategy> const kStrategy = [[BGRCachingImageFetchStrategy alloc] initWithFetchStrategy:[[BGRURLSessionImageFetchStrategy alloc] init]];
```
Comment (gershonho) [inline Baygur/BGRImageCell.mm]: 1. More than 100 characters in line.
2. Doc.

File: Baygur/BGRImageCell.mm [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+- (void)setup {
+//  self.autoresizingMask = UIViewAutoresizingNone;
```
Comment (gershonho) [inline Baygur/BGRImageCell.mm]: Drop

File: Baygur/BGRImageCell.mm [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+//  self.contentView.autoresizingMask = UIViewAutoresizingNone;
+  _imageView = [[BGRRemoteImageView alloc] init];
+  self.imageView.fetchStrategy = kStrategy; //[[BGRURLSessionImageFetchStrategy alloc] init];
```
Comment (gershonho) [inline Baygur/BGRImageCell.mm]: Drop comment.

File: Baygur/BGRImageDataSource.h [hunk 13]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) 2019 Lightricks. All rights reserved.
+// Created by Itai Shaked.
+
+NS_ASSUME_NONNULL_BEGIN
+
+@class BGRImageMetadata;
+
+@interface BGRImageDataSource : NSObject <UICollectionViewDataSource>
```
Comment (gershonho) [inline Baygur/BGRImageDataSource.h]: 1.Doc
2. Add empty line.

File: Baygur/BGRImageDataSource.h [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+- (instancetype)init NS_UNAVAILABLE;
+
+-(instancetype)initWithImages:(NSArray<BGRImageMetadata *> *)images
```
Comment (gershonho) [inline Baygur/BGRImageDataSource.h]: Doc

File: Baygur/BGRImageDataSource.h [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+              imageCellIdentifier:(NSString *)imageCellIdentifier
+        promotionalCellIdentifier:(NSString *)promotionalCellIdentifier NS_DESIGNATED_INITIALIZER;
+
+@property (readonly, nonatomic) NSArray<BGRImageMetadata *> *images;
```
Comment (gershonho) [inline Baygur/BGRImageDataSource.h]: Doc all properties, add empty lines

File: Baygur/BGRImageDataSource.mm [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+#import "BGRImageDataSource.h"
+
+#import "BGRImageDisplayer.h"
+#import "BGRPromoCell.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@implementation BGRImageDataSource
+
+- (instancetype)initWithImages:(NSArray<BGRImageMetadata *> *)images
```
Comment (gershonho) [inline Baygur/BGRImageDataSource.mm]: Indent

File: Baygur/BGRImageDataSource.mm [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+#pragma mark UICollectionViewDataSource
+#pragma mark -
+
+- (__kindof UICollectionViewCell *)collectionView:(UICollectionView *)collectionView
+                           cellForItemAtIndexPath:(NSIndexPath *)indexPath {
+  if ([self isIndexOfImage:indexPath.row]) {
+    auto imageCell = [collectionView dequeueReusableCellWithReuseIdentifier:self.imageCellIdentifier
+                                                               forIndexPath:indexPath];
+    LTAssert([imageCell conformsToProtocol:@protocol(BGRImageDisplayer)],
+        @"Cell does not implement BGRImageDisplayer.");
```
Comment (gershonho) [inline Baygur/BGRImageDataSource.mm]: indent

File: Baygur/BGRImageDataSource.mm [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+    [(id<BGRImageDisplayer>)imageCell
+        configureForImageMetadata:self.images[[self imageIndexForIndex:indexPath.row]]];
+    return imageCell;
+  } else {
+    BGRPromoCell *cell =
+        (BGRPromoCell *)[collectionView
+        dequeueReusableCellWithReuseIdentifier:self.promotionalCellIdentifier
+                         forIndexPath:indexPath];
+
+    //[cell showPromo];
```
Comment (gershonho) [inline Baygur/BGRImageDataSource.mm]: drop

File: Baygur/BGRImageDataSource.mm [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+        @"Cell does not implement BGRImageDisplayer.");
+    [(id<BGRImageDisplayer>)imageCell
+        configureForImageMetadata:self.images[[self imageIndexForIndex:indexPath.row]]];
+    return imageCell;
+  } else {
+    BGRPromoCell *cell =
+        (BGRPromoCell *)[collectionView
+        dequeueReusableCellWithReuseIdentifier:self.promotionalCellIdentifier
+                         forIndexPath:indexPath];
+
```
Comment (gershonho) [inline Baygur/BGRImageDataSource.mm]: drop

File: Baygur/BGRImageDataSource.mm [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+  if ([self isIndexOfImage:indexPath.row]) {
+    auto imageCell = [collectionView dequeueReusableCellWithReuseIdentifier:self.imageCellIdentifier
+                                                               forIndexPath:indexPath];
+    LTAssert([imageCell conformsToProtocol:@protocol(BGRImageDisplayer)],
+        @"Cell does not implement BGRImageDisplayer.");
+    [(id<BGRImageDisplayer>)imageCell
+        configureForImageMetadata:self.images[[self imageIndexForIndex:indexPath.row]]];
+    return imageCell;
+  } else {
+    BGRPromoCell *cell =
```
Comment (gershonho) [inline Baygur/BGRImageDataSource.mm]: 1. `auto`
2. indent

File: Baygur/BGRImageDisplayer.h [hunk 21]
```diff
@@ -0,0 +1,12 @@
+// Copyright (c) 2019 Lightricks. All rights reserved.
+// Created by Itai Shaked.
+
+NS_ASSUME_NONNULL_BEGIN
+
+@class BGRImageMetadata;
+
+@protocol BGRImageDisplayer <NSObject>
```
Comment (gershonho) [inline Baygur/BGRImageDisplayer.h]: doc

File: Baygur/BGRImageView.h [hunk 22]
```diff
@@ -6,8 +6,12 @@ NS_ASSUME_NONNULL_BEGIN
 /// View that wraps an image in a \c UIScrollView and loads images in the background.
 @interface BGRImageView : UIView
 
+- (void)setImage:(nullable UIImage *)image;
+
 /// URL of image to display. The image is loaded in the background. 
-@property (readwrite, nonatomic) NSURL *imageURL;
+//@property (readwrite, nonatomic) NSURL *imageURL;
```
Comment (gershonho) [inline Baygur/BGRImageView.h]: drop

File: Baygur/BGRImageView.h [hunk 23]
```diff
@@ -6,8 +6,12 @@ NS_ASSUME_NONNULL_BEGIN
...
+
+@property (readwrite, nonatomic) BOOL interactive;
```
Comment (gershonho) [inline Baygur/BGRImageView.h]: doc

**Comment by gershonho** [comment]
Round.

File: Baygur/BGRCachingImageFetchStrategy.h (hunk 1 above)
Comment (ltxitai) [reply Baygur/BGRCachingImageFetchStrategy.h]: Done.

File: Baygur/BGRCachingImageFetchStrategy.h (hunk 2 above)
Comment (ltxitai) [reply Baygur/BGRCachingImageFetchStrategy.h]: Done.

File: Baygur/BGRCachingImageFetchStrategy.mm (hunk 3 above)
Comment (ltxitai) [reply Baygur/BGRCachingImageFetchStrategy.mm]: Done.

File: Baygur/BGRRemoteImageFetchStrategy.h (hunk 4 above)
Comment (ltxitai) [reply Baygur/BGRRemoteImageFetchStrategy.h]: Done.

File: Baygur/BGRRemoteImageFetchStrategy.h (hunk 5 above)
Comment (ltxitai) [reply Baygur/BGRRemoteImageFetchStrategy.h]: Done.

File: Baygur/BGRImageView.mm (hunk 6 above)
Comment (ltxitai) [reply Baygur/BGRImageView.mm]: Done.

File: Baygur/BGRImageCell.h (hunk 7 above)
Comment (ltxitai) [reply Baygur/BGRImageCell.h]: Done.

File: Baygur/BGRImageCell.mm (hunk 8 above)
Comment (ltxitai) [reply Baygur/BGRImageCell.mm]: Done.

File: Baygur/BGRImageCell.mm (hunk 9 above)
Comment (ltxitai) [reply Baygur/BGRImageCell.mm]: Done.

File: Baygur/BGRImageCell.mm (hunk 10 above)
Comment (ltxitai) [reply Baygur/BGRImageCell.mm]: Done.

File: Baygur/BGRImageCell.mm (hunk 11 above)
Comment (ltxitai) [reply Baygur/BGRImageCell.mm]: Done.

File: Baygur/BGRImageCell.mm (hunk 12 above)
Comment (ltxitai) [reply Baygur/BGRImageCell.mm]: Done.

File: Baygur/BGRImageDataSource.h (hunk 13 above)
Comment (ltxitai) [reply Baygur/BGRImageDataSource.h]: Done.

File: Baygur/BGRImageDataSource.h (hunk 14 above)
Comment (ltxitai) [reply Baygur/BGRImageDataSource.h]: Done.

File: Baygur/BGRImageDataSource.h (hunk 15 above)
Comment (ltxitai) [reply Baygur/BGRImageDataSource.h]: Done.

File: Baygur/BGRImageDataSource.mm (hunk 16 above)
Comment (ltxitai) [reply Baygur/BGRImageDataSource.mm]: Done.

File: Baygur/BGRImageDataSource.mm (hunk 18 above)
Comment (ltxitai) [reply Baygur/BGRImageDataSource.mm]: Done.

File: Baygur/BGRImageDataSource.mm (hunk 19 above)
Comment (ltxitai) [reply Baygur/BGRImageDataSource.mm]: Done.

File: Baygur/BGRImageDataSource.mm (hunk 20 above)
Comment (ltxitai) [reply Baygur/BGRImageDataSource.mm]: Done.

File: Baygur/BGRImageDisplayer.h (hunk 21 above)
Comment (ltxitai) [reply Baygur/BGRImageDisplayer.h]: Done.

File: Baygur/BGRImageView.h (hunk 22 above)
Comment (ltxitai) [reply Baygur/BGRImageView.h]: Done.

File: Baygur/BGRImageView.h (hunk 23 above)
Comment (ltxitai) [reply Baygur/BGRImageView.h]: Done.

File: Baygur/BGRImageDataSource.mm (hunk 17 above)
Comment (ltxitai) [reply Baygur/BGRImageDataSource.mm]: Done.

**Comment by ltxitai** [comment]
Round. 
Squashed changes to a single commit, also added tests for `BGRCachingImageFetchStrategy`.

**Pull request closed without merging**